### PR TITLE
_cat/indices with Security, hide names when wildcard

### LINKED ISSUE #37190: 403 for no index permission from _cat/indices call
**Elasticsearch version** (`bin/elasticsearch --version`): 7.0.0-SNAPSHOT

**Description of the problem including expected versus actual behavior**:
When calling `_cat/indices` with no index list, indices for which the user does not have privileges simply omit the docs count and storage size information.  However, calling the same API _with_ an index for which the user does not have privileges (`_cat/indices/index_you_cant_see_info_for`) returns a 403 error:
```
{
  "error": {
    "root_cause": [
      {
        "type": "security_exception",
        "reason": "action [indices:monitor/stats] is unauthorized for user [billy]"
      }
    ],
    "type": "security_exception",
    "reason": "action [indices:monitor/stats] is unauthorized for user [billy]"
  },
  "status": 403
}
```

I think the API should behave the same in either case, just return the data without storage size and docs count.

### PR DESCRIPTION
**Pull request by albertzaharovits**
This changes the output of the `_cat/indices` API with `Security` enabled.

~Previously, it was~ It is possible to only display the index name (and possibly the index health, depending on the options) but not its stats (doc count, merges, size, etc). ~However, under usual circumstances, it was not possible to actually observe this, maybe just under a transient state (the index name should exist, but with no shards assigned).~  This is the case for closed indices which have index metadata in the cluster state but no associated shards, hence no shard stats. However, when `Security` is enabled, and the request contains wildcards, **open** indices without stats are a common occurrence. This is because the index names in the response table are picked up directly from the cluster state which is not filtered by `Security`'s _indexNameExpressionResolver_, unlike the
stats data which is populated by the indices stats API which does go through the index name resolver.
This is a bug, because it is circumventing `Security`'s function to hide unauthorized indices.

This has been fixed by displaying the index names as they are resolved by the indices stats API. The outputs of these two APIs is now very similar: same index names , similar data but different format.

Closes https://github.com/elastic/elasticsearch/issues/37190

_Even though the main code changes went into the server code, they impact the `Security` plugin, so I have labeled it in consequence._

cc @elastic/es-core-features

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java [hunk 1]
```diff
@@ -95,13 +95,8 @@ public RestChannelConsumer doCatRequest(final RestRequest request, final NodeCli
...
             public void processResponse(final ClusterStateResponse clusterStateResponse) {
-                final ClusterState state = clusterStateResponse.getState();
-                final Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, strictExpandIndicesOptions, indices);
-                // concreteIndices should contain exactly the indices in state.metaData() that were selected by clusterStateRequest using
-                // IndicesOptions.strictExpand(). We select the indices again here so that they can be displayed in the resulting table
-                // in the requesting order.
-                assert concreteIndices.length == state.metaData().getIndices().size();
-
+                final ClusterState clusterState = clusterStateResponse.getState();
+                final IndexMetaData[] indicesMetaData = getOrderedIndexMetaData(indices, clusterState, strictExpandIndicesOptions);
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java]: As the code suggests, `indexNameExpressionResolver.concreteIndices` should probably return the resolved indices expression in the order that they were requested. It does not appear to be the case now, but I nevertheless kept the logic based on this assumption in the spirit of introducing a single change in a PR.

File: server/src/test/java/org/elasticsearch/rest/action/cat/RestIndicesActionTests.java [hunk 2]
```diff
@@ -170,4 +188,19 @@ private IndicesStatsResponse randomIndicesStatsResponse(final Index[] indices) {
             shardStats.toArray(new ShardStats[shardStats.size()]), shardStats.size(), shardStats.size(), 0, emptyList()
         );
     }
+
+    private <T> T[] removeRandomElement(T[] array) {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/rest/action/cat/RestIndicesActionTests.java]: maybe you can use a List and avoid all the ceremony need to remove an element from an array?

File: server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/rest/action/cat/RestIndicesAction.java]: we put indices in a set I think, then ordering is not deterministic

**Review by javanna — APPROVED** [review]

**Comment by albertzaharovits** [comment]
@javanna FYI after checking on the test failures I realized we need a special case for closed indices.
These have metadata but no shards, hence no shard stats, so they were falling in the "unauthorized" indices case, but they have to be displayed. I have pushed https://github.com/elastic/elasticsearch/pull/38824/commits/137ad57451aab4b1635a425f92a6e37ae1aee632 .

**Comment by javanna** [comment]
sounds good @albertzaharovits thanks for pinging

**Comment by rhass** [comment]
You can still view the index names when hitting the  `/_cat/shards` API endpoint.

**Pull request merged**